packages/upgrade/src/actions/verify.ts

```diff
@@ -0,0 +1,153 @@
+
+
+export async function verify(
+	ctx: Pick<Context, 'version' | 'packages' | 'cwd' | 'dryRun' | 'exit'>
+) {
+	const registry = await getRegistry();
+
+	if (!ctx.dryRun) {
+		const online = await isOnline(registry);
+		if (!online) {
+			bannerAbort();
+			log('');
+			error('error', `Unable to connect to the internet.`);
+			ctx.exit(1);
+		}
+	}
+
+	await verifyAstroProject(ctx);
+
+	const ok = await verifyVersions(ctx, registry);
+	if (!ok) {
+		bannerAbort();
+		log('');
+		error('error', `Version ${color.reset(ctx.version)} ${color.dim('could not be found!')}`);
+		await info('check', 'https://github.com/withastro/astro/releases');
+		ctx.exit(1);
+	}
+}
+
+function isOnline(registry: string): Promise<boolean> {
+	const { host } = new URL(registry);
+	return dns.lookup(host).then(
+		() => true,
+		() => false
+	);
+}
+
+function safeJSONParse(value: string) {
+	try {
+		return JSON.parse(value);
+	} catch {}
+	return {}
+}
+
+async function verifyAstroProject(ctx: Pick<Context, 'cwd' | 'version' | 'packages'>) {
+	const packageJson = new URL('./package.json', ctx.cwd);
+    if (!existsSync(packageJson)) return false;
+	const contents = await readFile(packageJson, { encoding: 'utf-8' });
+	if (!contents.includes('astro')) return false;
+
+	const { dependencies = {}, devDependencies = {} } = safeJSONParse(contents)
+	if (dependencies['astro'] === undefined && devDependencies['astro'] === undefined) return false;
+	
+	// Side-effect! Persist dependency info to the shared context
+	collectPackageInfo(ctx, dependencies, devDependencies);
+	
+	return true;
+}
+
+function isAstroPackage(name: string) {
+	return name === 'astro' || name.startsWith('@astrojs/');
+}
+
+function collectPackageInfo(ctx: Pick<Context, 'version' | 'packages'>, dependencies: Record<string, string>, devDependencies: Record<string, string>) {
+	for (const [name, currentVersion] of Object.entries(dependencies)) {
+		if (!isAstroPackage(name)) continue;
+		ctx.packages.push({
+			name,
+			currentVersion,
+			targetVersion: ctx.version,
+		})
+	}
+	for (const [name, currentVersion] of Object.entries(devDependencies)) {
+		if (!isAstroPackage(name)) continue;
+		ctx.packages.push({
+			name,
+			currentVersion,
+			targetVersion: ctx.version,
+			isDevDependency: true
+		})
+	}
+}
+
+async function verifyVersions(ctx: Pick<Context, 'version' | 'packages'>, registry: string) {
+	const tasks: Promise<void>[] = [];
+	for (const packageInfo of ctx.packages) {
+		tasks.push(resolveTargetVersion(packageInfo, registry));
+	}
+	await Promise.all(tasks);
+	for (const packageInfo of ctx.packages) {
+		if (!packageInfo.targetVersion) {
+			return false;
+		}
+	}
+	return true;
+}
+
+async function resolveTargetVersion(packageInfo: PackageInfo, registry: string): Promise<void> {
+	const packageMetadata = await fetch(`${registry}/${packageInfo.name}`, { headers: { accept: 'application/vnd.npm.install-v1+json' }});
+	const { "dist-tags": distTags } = await packageMetadata.json();
```

Comment:
These two operations can fail, we should catch the error somewhere.

If we decide to handle the error inside the caller, we should call this function `tryResolveTargetVersion`.

Also, the `fetch` call can return a non-200 code, and we should check it first.

Reply:
Good call. Refactored this to handle `>= 400` status codes in [203530f](https://github.com/withastro/astro/pull/8525/commits/203530febef74e9610f80f8cc484d45d6304ed71) and made sure that our existing error message was triggered if these calls fail.